### ISSUE TITLE
Issue with bootstrapping Tor when the forward proxy connects ipv6

### ISSUE THREAD

**Issue by Buddle** [nghttp2/nghttp2#1059]
**Tor  <<>> nghttpx(forward proxy) << Internet >> nghttpx <<>> proxy**
nghttpx is the https/http proxy for Tor and Tor access the Internet through it.

If the proxy is connected to the server with IPv6, Tor fails to finish bootstrap and sticks at 50% . Tor is able to communicate with its directory server and get the relay addresses. But Tor can't establish Tor circuits.

If the proxy is connected to the server with IPv4, Tor is able to establish the circuit and afterwards nghttpx restarts with IPv6 connections, Tor still works as normal since it doesn't rebuild circuits.(I am able to access the .onion sites). 
That means this issue only affects Tor's building connections to relay nodes.

This fragment and similitude  appear both in server and client log. 
> "CONNECT 178.62.66.18:9001 HTTP/2" 200 3373 "-" "-"
> nghttpx[1739]: [INFO] [DOWNSTREAM:0x7f72d35f1900] Deleting (shrpx_downstream.cc:166) nghttpx[1739]: [INFO] [DOWNSTREAM:0x7f72d35f1900] Deleted (shrpx_downstream.cc:196)
> nghttpx[1739]: [INFO] [DCONN:0x7f72d3628980] Time out (shrpx_http_downstream_connection.cc:57)
> nghttpx[1739]: [INFO] [DCONN:0x7f72d3628980] Timeout (shrpx_http2_upstream.cc:1307)
> nghttpx[1739]: [INFO] [DCONN:0x7f72d3628980] Note: this is tunnel connection (shrpx_http2_upstream.cc:1310)
> nghttpx[1739]: [INFO] [DCONN:0x7f72d3628980] Deleted (shrpx_http_downstream_connection.cc:223)
> nghttpx[1739]: [INFO] [DOWNSTREAM:0x7f72d35f2b00] HTTP response completed (shrpx_http2_upstream.cc:1795)
> nghttpx[1739]: [INFO] [UPSTREAM:0x7f72d35ca8c0] Send RST_STREAM to tunneled stream stream_id=1 to finish off incomplete request (shrpx_http2_upstream.cc:647)
> nghttpx[1739]: [INFO] [UPSTREAM:0x7f72d35ca8c0] RST_STREAM stream_id=1 with error_code=0 (shrpx_http2_upstream.cc:1354)
> nghttpx[1739]: [INFO] [UPSTREAM:0x7f72d35ca8c0] Stream stream_id=1 is being closed (shrpx_http2_upstream.cc:63)
>

**Comment by Buddle** [comment]
Tor stuck for 10 minutes and stop bootstrapping.

> raspberrypi Tor[838]: Proxy Client: unable to connect to 93.115.91.66:443
> raspberrypi Tor[838]: The https proxy sent back an unexpected status code 500 ("Internal Server Error"). Closing.
> raspberrypi Tor[838]: Proxy Client: unable to connect to 93.115.91.66:443
> raspberrypi Tor[838]: The https proxy sent back an unexpected status code 500 ("Internal Server Error"). Closing.
> raspberrypi Tor[838]: Proxy Client: unable to connect to 93.115.91.66:443

**Comment by tatsuhiro-t** [comment]
I'm not familiar with Tor architecture, I'd like to know the situation.

> Tor <<>> nghttpx-A(forward proxy) << Internet >> nghttpx-B <<>> proxy

So if I understand correctly, nghttpx-A is configured without -s (--http2-proxy) option, and nghttpx-B is configured with -s (--http2-proxy).  And proxy is standard HTTP proxy, just like squid.

You have a problem when Tor establishing connections using IPv6.  Is that right?  But the logs you pasted above does not include any IPv6 addresses.

Because the actual connection to IPv6 host is done by final *proxy* (e.g., squid), if nghttpx correctly issues CONNECT request to the proxy, it might be an issue of the final proxy.

**Comment by Buddle** [comment]
Sorry for my poor English
> With the --http2-proxy option, it works as forward proxy, and it is so called secure HTTP/2 proxy (aka SPDY proxy):
> Client <-- (HTTP/2, SPDY, HTTP/1.1) --> nghttpx <-- (HTTP/1.1) --> Proxy

Actually A is the client and B is the nghttpx.

A is on my local device and B is on remote server which has both IPv4 and IPv6 address. Both of them are configured with http2-proxy. A could be Chrome or firefox but Tor doesn't support http2 proxy, so another nghttpx "A" works as client.

Configuration is the same as
https://github.com/nghttp2/nghttp2/issues/240

The "IPv6" means how A connects B, the backend of A configuration, which is bound to the domain of B. If IPv4 address is resolved, Tor bootstrap works. On the contrary, if AAAA record is given, Tor fails to bootstrap. Tor's cache dir is on tmpfs, being reserved except when system reboot. Tor stores bootstrap information into its cache dir for the next time it runs. If nghttpx A restarts with IPv6 address as backend, Tor manages to bootstrap **only when bootstrap cache is handled**. (That's why I found this issue affect only bootstrap. )

**Comment by Buddle** [comment]
> Tor bootstrap refers to the process of attempting to connect to the Tor network (successfully or unsuccessfully).
 
By whonix

> If nghttpx A restarts with IPv6 address as backend, Tor manages to bootstrap only when bootstrap cache is handled.

Tor network could be accessed only after Tor finishes bootstrap. A restarted Tor doesn't need re-bootstrap because of its cache and I am able to connect to its network with either IPv4 or IPv6  backend.

**Comment by tatsuhiro-t** [comment]
Thank you for clarification.   I understand that the problem occurs when A is connecting to B with IPv6 address.  Will look into it.

**Comment by tatsuhiro-t** [comment]
Does nghttpx listen to IPv6 address?

**Comment by Buddle** [comment]
I set
`frontend=*,12345`

**Comment by tatsuhiro-t** [comment]
Could you check that `ss -tanl|grep 12345` actually indicates that nghttpx listens to IPv6 address?
Can you send request to nghttpx B directly using IPv6 address with, say, curl?

**Comment by Buddle** [comment]
Nghttpx B is really listening to IPv6 address.
I can also communicate with nghttpx B with Chrome directly and IPv6 address was revealed in Chrome network panel.  I set two domains, either of which only has one record, A or AAA and neither of them would be given simultaneously. So I am sure when is IPv6 is on.

**Comment by tatsuhiro-t** [comment]
So far I could not find any issue regarding IPv6 connectivity between nghttpx proxies.
I uses --backend-address-family=ipv6 to force IPv6 for backend address, and have no issues at all.

**Comment by Buddle** [comment]
`backend-address-family` didn't help.
Does nghttpx A (client) need `http2-proxy=yes`?

At least  Chrome has no issue regarding IPv6 connectivity between nghttpx proxies ( "https proxy" direct to the server or "proxy" just to nghttpx A) but Tor bootstrap.

Do you set 
`HTTPProxy nhgttpxA:Frontport`
`HTTPSProxy nhgttpxA:Frontport`
in Tor's torrc?

**Comment by tatsuhiro-t** [comment]
> Does nghttpx A (client) need http2-proxy=yes?

I think yes.  Initially, I didn't but it is probably better to set it.

> backend-address-family didn't help.

If the cause is purely IPv6, backend-address-family=IPv4 should solve this problem.  If it doesn't then it suggests that something other may involve in this issue.

**Comment by Buddle** [comment]
> If the cause is purely IPv6, backend-address-family=IPv4 should solve this problem. If it doesn't then it suggests that something other may involve in this issue.

I have said there was no issue regarding IPv4 backend.

https://bridges.torproject.org/
You can get bridges to test. I just added three, Tor debug log

> Tor[1818]: connection_read_proxy_handshake(): enter state PROXY_HTTPS_WANT_CONNECT_OK
Tor[1818]: fetch_from_buf_http(): headerlen 107, bodylen 545.
Tor[1818]: The https proxy sent back an unexpected status code 500 ("Internal Server Error"). Closing.
Tor[1818]: connection_read_proxy_handshake(): leaving state PROXY_HTTPS_WANT_CONNECT_OK
Tor[1818]: Proxy Client: unable to connect to 185.100.84.187:57092

This would let Tor bootstrap connect to bridges as soon as Tor connects to its  dir server.

**Comment by Buddle** [comment]
By the way, nghttpx B ,the server side could get the request from Tor.

**Comment by Buddle** [comment]
> Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b800] SSL/TLS handshake completed (shrpx_client_handler.cc:197)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b800] The negotiated next protocol: h2 (shrpx_client_handler.cc:583)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca280] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca280] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 109.213.205.211:42035#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Attaching to DOWNSTREAM:0x7f07905f2500 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] HTTP request headers. stream_id=1#012CONNECT 109.213.205.211:42035 HTTP/1.1#015#012Host: 109.213.205.211:42035#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 184.0.63.5:443#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Attaching to DOWNSTREAM:0x7f07905f2800 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] HTTP request headers. stream_id=1#012CONNECT 184.0.63.5:443 HTTP/1.1#015#012Host: 184.0.63.5:443#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:55:55  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 185.100.84.187:57092#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b800] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:55:55  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b800] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Attaching to DOWNSTREAM:0x7f07905f2b00 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] HTTP request headers. stream_id=1#012CONNECT 185.100.84.187:57092 HTTP/1.1#015#012Host: 185.100.84.187:57092#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:55:55  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] HTTP response header completed (shrpx_http2_upstream.cc:1601)
Nov 18 08:55:56  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] HTTP response headers. stream_id=1#012:status: 500#012content-type: text/html#012server: tinyproxy/1.8.4#012 (shrpx_http2_upstream.cc:1919)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] EOF. stream_id=1 (shrpx_http2_upstream.cc:1267)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Deleted (shrpx_http_downstream_connection.cc:223)
Nov 18 08:55:56  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] Downstream body was ended by EOF (shrpx_http2_upstream.cc:1279)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] HTTP response completed (shrpx_http2_upstream.cc:1795)
Nov 18 08:55:56  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] Send RST_STREAM to stream stream_id=1 to finish off incomplete request (shrpx_http2_upstream.cc:647)
Nov 18 08:55:56  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] RST_STREAM stream_id=1 with error_code=0 (shrpx_http2_upstream.cc:1354)
Nov 18 08:55:56  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca780] Stream stream_id=1 is being closed (shrpx_http2_upstream.cc:63)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] Deleting (shrpx_downstream.cc:166)
Nov 18 08:55:56  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] Deleted (shrpx_downstream.cc:196)

Here is the log when Tor was trying to connect the bridges

**Comment by Buddle** [comment]
>Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053c000] SSL/TLS handshake completed (shrpx_client_handler.cc:197)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053c000] The negotiated next protocol: h2 (shrpx_client_handler.cc:583)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca140] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca140] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 184.0.63.5:443#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053c000] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053c000] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Attaching to DOWNSTREAM:0x7f07905f1c00 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] HTTP request headers. stream_id=1#012CONNECT 184.0.63.5:443 HTTP/1.1#015#012Host: 184.0.63.5:443#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628280] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:02:24  nghttpx[959]: [INFO] SSL/TLS handshake completed (shrpx_connection.cc:507)
Nov 18 08:02:24  nghttpx[959]: [INFO] cipher=ECDHE-ECDSA-AES256-GCM-SHA384 protocol=TLSv1.2 resumption=yes session_id=8af688e35bf556adbbb97634ac41ea8ff198860908d81f34a9ea06d8e1c36c5c (shrpx_connection.cc:510)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] SSL/TLS handshake completed (shrpx_client_handler.cc:197)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] The negotiated next protocol: h2 (shrpx_client_handler.cc:583)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca3c0] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca3c0] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 109.213.205.211:42035#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053a800] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Attaching to DOWNSTREAM:0x7f07905f2800 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] HTTP request headers. stream_id=1#012CONNECT 109.213.205.211:42035 HTTP/1.1#015#012Host: 109.213.205.211:42035#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628600] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:02:24  nghttpx[959]: [INFO] SSL/TLS handshake completed (shrpx_connection.cc:507)
Nov 18 08:02:24  nghttpx[959]: [INFO] cipher=ECDHE-ECDSA-AES256-GCM-SHA384 protocol=TLSv1.2 resumption=yes session_id=0d3c41aae9a34f0fc8f4d627a98f5ddb265df22e11e74b0c21f41a92ba76c834 (shrpx_connection.cc:510)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] SSL/TLS handshake completed (shrpx_client_handler.cc:197)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] The negotiated next protocol: h2 (shrpx_client_handler.cc:583)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] Received upstream request HEADERS stream_id=1 (shrpx_http2_upstream.cc:270)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] HTTP request headers. stream_id=1#012:method: CONNECT#012:authority: 185.100.84.187:57092#012 (shrpx_http2_upstream.cc:326)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] Downstream address group_idx: 0 (shrpx_client_handler.cc:1000)
Nov 18 08:02:24  nghttpx[959]: [INFO] [CLIENT_HANDLER:0x7f079053b000] Downstream connection pool is empty. Create new one (shrpx_client_handler.cc:1119)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Attaching to DOWNSTREAM:0x7f07905f2b00 (shrpx_http_downstream_connection.cc:236)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Connecting to downstream server (shrpx_http_downstream_connection.cc:417)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] HTTP request headers. stream_id=1#012CONNECT 185.100.84.187:57092 HTTP/1.1#015#012Host: 185.100.84.187:57092#015#012#015#012 (shrpx_http_downstream_connection.cc:694)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Connected to downstream host (shrpx_http_downstream_connection.cc:1433)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] HTTP response header completed (shrpx_http2_upstream.cc:1601)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] HTTP response headers. stream_id=1#012:status: 500#012content-type: text/html#012server#012 (shrpx_http2_upstream.cc:1919)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] EOF. stream_id=1 (shrpx_http2_upstream.cc:1267)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DCONN:0x7f0790628980] Deleted (shrpx_http_downstream_connection.cc:223)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] Downstream body was ended by EOF (shrpx_http2_upstream.cc:1279)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] HTTP response completed (shrpx_http2_upstream.cc:1795)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] Send RST_STREAM to stream stream_id=1 to finish off incomplete request (shrpx_http2_upstream.cc:647)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] RST_STREAM stream_id=1 with error_code=0 (shrpx_http2_upstream.cc:1354)
Nov 18 08:02:24  nghttpx[959]: [INFO] [UPSTREAM:0x7f07905ca500] Stream stream_id=1 is being closed (shrpx_http2_upstream.cc:63)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] Deleting (shrpx_downstream.cc:166)
Nov 18 08:02:24  nghttpx[959]: [INFO] [DOWNSTREAM:0x7f07905f2b00] Deleted (shrpx_downstream.cc:196)

Above is the log when Tor was trying to connects the bridges

**Comment by tatsuhiro-t** [comment]
The log says that a client tried to connect to 185.100.84.187:57092, but it got HTTP 500 error.
nghttpx B log says that it got CONNECT request from nghttpx A, and forwarded to backend proxy, and it got HTTP 500 error.

What I haven't understand is that how IPv6 between nghttpx A and B affects the above interaction.
The log shows that nghttpx A successfully forwarded request to B, which means IPv6 works (otherwise how nghttpx received the request in the first place?).
I suspect that there are other errors related to IPv6, but I'm not sure.

I'm not familiar to Tor, so I don't know how to use bridge to debug this issue.  Could you show me the step-by-step reproduction procedure?

**Comment by Buddle** [comment]
# Get Tor Bridge (Not necessary)
https://bridges.torproject.org/
Just click "get bridges" and then "just give me bridges". Paste the three given address and their fingerprint into torrc
Just like
`UseBridges 1`
`Bridge 184.0.63.5:443 263423EBD3643CEC5577474A0F23460F4BD7B8B3`
`Bridge 185.100.84.187:57092 158A7DD6EDCA9041D7FA6ED7593E7BB43F82BD84`
`Bridge 109.213.205.211:42035 309FFCA26AC94110EC5E6E408A9EC8490D41A1C0`

# Tor configuration

>#8080 is the port nghttpx A listening
DataDirectory /var/run/tor
#you can set what you want. It is cache I mentioned.
HTTPProxy 127.0.0.1:8080
HTTPSProxy 127.0.0.1:8080
AvoidDiskWrites 1
ClientOnly 1
DirReqStatistics 0
#192.168.1.1 is the device running nghttpx A
SOCKSPort 192.168.1.1:9150
Log notice syslog
#UseBridges 1
#Bridge 184.0.63.5:443 263423EBD3643CEC5577474A0F23460F4BD7B8B3
#Bridge 185.100.84.187:57092 158A7DD6EDCA9041D7FA6ED7593E7BB43F82BD84
#Bridge 109.213.205.211:42035 309FFCA26AC94110EC5E6E408A9EC8490D41A1C0


# nghttpx A with IPv6 backend
Make sure the cache dir is empty.Then start tor. This issue occur. Tor fails to bootstrap and Tor network couldn't be connected to.

# Why bootstrap
1. Stop Tor.
2. **Cache dir must be cleaned by hand** after Tor has been stopped.
3. Restart nghttpx A with IPv4 backend.
4. Start Tor again
You would see Tor finish bootstrap in syslog. And Chrome can access Tor network via socks5 port.

The major procedure is to clean cache to make sure Tor connects bridges again rather than bootstrap from cached consensus.

**Comment by tatsuhiro-t** [comment]
Thank you.  I still don't know how to use these configurations.
What program should I download for this purpose?  Tor browser?
Note I completely don't know about any Tor related stuff.
Where is the cache directory in the local file system?

**Comment by Buddle** [comment]
I just installed Tor from "apt-get install tor", not Tor Browser. The other computers can access Tor network with shared Tor's "SOCKSPort" of the device running Tor. You can run tor by hand.

> /usr/bin/tor **--defaults-torrc /usr/share/tor/tor-service-defaults-torrc** -f **/etc/tor/torrc** --RunAsDaemon 0

/etc/tor/torrc contains custom parameters overriding the default. The cache directory varies in different OS. So its location should be defined in /etc/tor/torrc by "DataDirectory".

**Comment by tatsuhiro-t** [comment]
It works for me.  IPv6 just works.

**Comment by Buddle** [comment]
Thank you for test.
I really feel puzzled now why nghttpx B received the requests from IPv6 network and responded with **RST_STREAM**

**Comment by Buddle** [comment]
@tatsuhiro-t 

There is a new issue in such situation that attachments are not able to be uploaded on Gmail website.
Could you reproduce ?

Chrome shows net::ERR_CONNECTION_RESET and net::ERR_SPDY_PING_FAILED.
Log

> nghttpx[992]: [INFO] [DCONN:0x71849020] Submit RST_STREAM for DOWNSTREAM:0x7180d580, stream_id=5, error_code=2 (shrpx_http2_downstream_connection.cc:151)
